crates/interpreter/src/valid.rs

```diff
@@ -503,29 +511,58 @@ impl<'a, 'm> Expr<'a, 'm> {
         match instr {
             Unreachable => self.stack_polymorphic(),
             Nop => (),
-            Block(b) => self.push_label(self.blocktype(&b)?, LabelKind::Block)?,
-            Loop(b) => self.push_label(self.blocktype(&b)?, LabelKind::Loop)?,
+            Block(b) => {
+                self.push_label(self.blocktype(&b)?, LabelKind::Block, vec![])?;
+                self.side_table.push(None);
+            }
+            Loop(b) => {
+                self.push_label(
+                    self.blocktype(&b)?,
+                    // TODO(dev/fast-interp): Store the parser position in the LabelKind (as done
+                    // in exec.rs).
+                    LabelKind::Loop,
+                    vec![],
+                )?;
+                self.side_table.push(None);
+            }
             If(b) => {
                 self.pop_check(ValType::I32)?;
-                self.push_label(self.blocktype(&b)?, LabelKind::If)?;
+                self.push_label(self.blocktype(&b)?, LabelKind::If(self.branch()), vec![])?;
+                self.push_entry_only();
             }
             Else => {
+                let LabelKind::If(SideTableBranch { parser, side_table_len }) = self.label().kind
+                else {
+                    unreachable!()
+                };
+                self.stitch(&SideTableBranch { parser, side_table_len });
+
                 let label = self.label();
-                check(core::mem::replace(&mut label.kind, LabelKind::Block) == LabelKind::If)?;
+                check(
+                    core::mem::replace(&mut label.kind, LabelKind::Block)
+                        == LabelKind::If(SideTableBranch { parser, side_table_len }),
+                )?;
                 let FuncType { params, results } = label.type_;
                 self.pops(results)?;
                 check(self.stack().is_empty())?;
                 self.label().polymorphic = false;
                 self.pushs(params);
+
+                let parser = self.parser.save();
+                let side_table_len = self.side_table.len();
+                self.label().branches.push(SideTableBranch { parser, side_table_len });
+                self.push_entry_only();
             }
             End => unreachable!(),
             Br(l) => {
-                self.pops(self.br_label(l)?)?;
+                let res = self.br_label(l)?;
+                self.pops(res)?;
```

Comment:
Interesting, is the borrow checker newly complaining about this?

Reply:
Yes, because `br_label()` now takes `&mut self` instead.